src/base/os.lua

```diff
@@ -159,7 +159,16 @@
 		elseif type(headerdirs) == "table" then
 			userpaths = headerdirs
 		end
-		paths = table.join(userpaths, paths)
+
+		for _, userpath in ipairs(userpaths) do
+			if path.isabsolute(userpath) then
+				paths = table.join({userpath}, paths)
+			else
+				for _, p in ipairs(paths) do
+					paths = table.join({path.join(p, userpath)}, paths)
```

Comment:
Same here

Reply:
I think the search should prioritize subdirectories. For example, if a user's computer has both `/usr/include/lua.h` and `/usr/include/lua5.1/lua.h`, and the user calls `os.findheader("lua.h", "lua5.1")`, they are more likely looking for the latter.

Reply:
In addition you iterate and modify `paths`, so `os.findheader('lib.h', {'dir1', dir2})` will result in something like
`{'/usr/dir1/dir2', '/usr/include/dir2', '/usr/dir1', '/usr/include'}`

A `table.transform(userpaths` to transform relative paths into a table, and keep absolute path as-is would be better, then flatten userpaths, and finally `paths = table.join(userpaths, paths)`